### PR TITLE
Added named tensor

### PR DESCRIPTION
**Pull request by junjihashimoto**
I implemented 3 different tensors to define named-tensor as follows.

- This PR with `Unnamed`-typeclass.
- https://github.com/hasktorch/hasktorch/pull/537
- https://github.com/hasktorch/hasktorch/pull/524

The method other than this PR is a lot of modification of the existing code.
When considering a way to gradually update an existing implementation to a new tensor (Named Tensor), it is better to put a layer that abstracts the `Unnamed Tensor` instead of modifying the definition of `Tensor` directly. 
So I'm making `Unnamed-typeclass` as the layer.
In addition to this, we can use multiple Tensor definitions by using that layer.

I'm thinking broadcast for the namedtensor.

### CODE REVIEW

File: hasktorch/src/Torch/Typed/Functional.hs [hunk 1]
```diff
@@ -654,13 +654,15 @@ pow exponent input = unsafePerformIO $ ATen.cast2 ATen.Managed.pow_tt input expo
...
+   Unnamed t, device ~ (UTDevice t), dtype ~ (UTDType t), shape ~ (UTShape t)
+  ) =>
   -- | input
-  Tensor device dtype shape ->
+  t ->
   -- | output
-  Tensor device dtype shape
-relu input = unsafePerformIO $ ATen.cast1 ATen.Managed.relu_t input
+  t
+relu input = unWrap $ unsafePerformIO $ ATen.cast1 ATen.Managed.relu_t (Wrap input)
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Typed/Functional.hs:665]: This modification does not effect current implementation like static-mnist-cnn of example-directory.

File: hasktorch/src/Torch/Typed/Factories.hs [hunk 2]
```diff
@@ -36,6 +36,18 @@ import Torch.Typed.Aux
 import Torch.Typed.Tensor
 import Prelude hiding (sin)
 
+instance Semigroup (Tensor device dtype shape) where
+  (<>) a b = UnsafeMkTensor $ D.add (toDynamic a) (toDynamic b)
+
+instance (TensorOptions shape dtype device) => Monoid (Tensor device dtype shape) where
```
Comment (tscholak) [inline hasktorch/src/Torch/Typed/Factories.hs:42]: why do you need this concrete monoid for Tensor and NamedTensor?
Perhaps consider https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Monoid.html#v:Sum or https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Monoid.html#v:Product from base.

File: hasktorch/src/Torch/Typed/Tensor.hs [hunk 3]
```diff
@@ -599,3 +665,43 @@ toDouble t = D.asValue . toDynamic . toCPU $ t
...
 toBool t = D.asValue . toDynamic . toCPU $ t
+
+--------------------------------------------------------------------------------
+-- NamedTensor
+--------------------------------------------------------------------------------
+
+data NamedTensor (device :: (D.DeviceType, Nat)) (dtype :: D.DType) (shape :: Shape) where
+  FromTensor :: forall device dtype shape' shape. shape ~ ToNats shape' => Tensor device dtype shape ->  NamedTensor device dtype shape'
+
+instance Unnamed (NamedTensor device dtype shape) where
```
Comment (tscholak) [inline hasktorch/src/Torch/Typed/Tensor.hs:676]: do you have an instance for `Tensor device dtype shape`, too?

File: hasktorch/src/Torch/Typed/Tensor.hs [hunk 4]
```diff
@@ -599,3 +665,43 @@ toDouble t = D.asValue . toDynamic . toCPU $ t
...
+  type UTShape (NamedTensor device dtype shape) = ToNats shape
+  type UTDevice (NamedTensor device dtype shape) = device
+  type UTDType (NamedTensor device dtype shape) = dtype
+  toUnnamed (FromTensor t) = t
+  fromUnnamed = FromTensor
+  toDynamic (FromTensor (UnsafeMkTensor t)) = t
+
+instance (KnownDevice device) => Num (NamedTensor device dtype shape) where
```
Comment (tscholak) [inline hasktorch/src/Torch/Typed/Tensor.hs:684]: you may be able to use the `Num` instance for `Tensor device dtype shape` here

**Review by tscholak — APPROVED** [review]
LGTM, well done!

File: hasktorch/src/Torch/Typed/Tensor.hs (hunk 3 above)
Comment (junjihashimoto) [reply hasktorch/src/Torch/Typed/Tensor.hs:676]: Yes. [This code](https://github.com/hasktorch/hasktorch/pull/536/files/2ab534287bd778c9325ead060f1a878fde0453da#diff-5757ffa202fb7ecd16a6aa1c8217948a432d109d10ae364fdb8963f6e2303d80R162) is the instance.

File: hasktorch/src/Torch/Typed/Factories.hs (hunk 2 above)
Comment (junjihashimoto) [reply hasktorch/src/Torch/Typed/Factories.hs:42]: I need a default constructor like zeros for this code.
https://github.com/hasktorch/hasktorch/pull/536/files/2ab534287bd778c9325ead060f1a878fde0453da#diff-ed8556e187a95c50a009496221d2e54ca9fe54ceafd52bbabd8470f10c1677fbR69-R79

Is it better to have an instance of Data.Default?

File: hasktorch/src/Torch/Typed/Tensor.hs [hunk 5]
```diff
@@ -682,21 +683,21 @@ instance Unnamed (NamedTensor device dtype shape) where
...
-  abs t = fromUnnamed $ UnsafeMkTensor $ abs $ toDynamic t
-  signum t = fromUnnamed $ UnsafeMkTensor $ signum $ toDynamic t
-  fromInteger i = fromUnnamed $ UnsafeMkTensor . D.toDevice (deviceVal @device) . D.asTensor @Int $ fromInteger @Int i
+  (+) a b = fromUnnamed $ toUnnamed a + toUnnamed b
+  (-) a b = fromUnnamed $ toUnnamed a - toUnnamed b
+  (*) a b = fromUnnamed $ toUnnamed a * toUnnamed b
+  negate = fromUnnamed . negate . toUnnamed
+  abs = fromUnnamed . abs . toUnnamed
+  signum = fromUnnamed . signum . toUnnamed
+  fromInteger = fromUnnamed . fromInteger
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Typed/Tensor.hs]: I've tried using an instance of Num. It looks good.

**Comment by junjihashimoto** [comment]
@tscholak 
Thank you for reviewing!

**Pull request merged**